pkgs/development/python-modules/torch/default.nix

```diff
@@ -307,6 +308,7 @@ in buildPythonPackage rec {
       cuda_nvrtc.lib
       cuda_nvtx.dev
       cuda_nvtx.lib # -llibNVToolsExt
+      cudatoolkit
```

Comment:
The premise of the redist packages (`cuda_*`) is that we can avoid dependence on `cudatoolkit` which is heavyweight and targeted for eventual deprecation. Let's add only the redist packages that are necessary to build here.

Reply:
Ok I see. However, the build fails as following:
```
-- Found CUDA: /nix/store/6v23vkbm2c9c3hihzlj3d2jkgsxrl8rm-cuda_nvcc-11.8.89 (found version "11.8")
-- The CUDA compiler identification is NVIDIA 11.8.89
-- Detecting CUDA compiler ABI info
-- Detecting CUDA compiler ABI info - done
-- Check for working CUDA compiler: /nix/store/6v23vkbm2c9c3hihzlj3d2jkgsxrl8rm-cuda_nvcc-11.8.89/bin/nvcc - skipped
-- Detecting CUDA compile features
-- Detecting CUDA compile features - done
-- Unable to find cuda_runtime.h in "/nix/store/6v23vkbm2c9c3hihzlj3d2jkgsxrl8rm-cuda_nvcc-11.8.89/include" for CUDAToolkit_INCLUDE_DIR.
-- Unable to find cublas_v2.h in either "" or "/nix/math_libs/include"
CMake Error at /nix/store/yi235g10sp8jx939zpfli0s74154ph3v-cmake-3.26.4/share/cmake-3.26/Modules/FindPackageHandleStandardArgs.cmake:230 (message):
  Could NOT find CUDAToolkit (missing: CUDAToolkit_INCLUDE_DIR) (found
  version "11.8.89")
Call Stack (most recent call first):
  /nix/store/yi235g10sp8jx939zpfli0s74154ph3v-cmake-3.26.4/share/cmake-3.26/Modules/FindPackageHandleStandardArgs.cmake:600 (_FPHSA_FAILURE_MESSAGE)
  cmake/Modules/FindCUDAToolkit.cmake:837 (find_package_handle_standard_args)
  cmake/public/cuda.cmake:59 (find_package)
  cmake/Dependencies.cmake:44 (include)
  CMakeLists.txt:722 (include)
```

It is not clear which specific library is missing...

Reply:
Hmm @SomeoneSerge has previously accomplished some wizardry-level feats getting CMake to play nicely with redist packages. Perhaps he might have some thoughts here?

Reply:
We've recently merged https://github.com/NixOS/nixpkgs/pull/249259 which made torch work without the runfile-based `cudatoolkit`, and at that point cuda_cudart discovery did work. I'm not yet sure what could be the cause this is breaking now with the update

Reply:
I tried rebasing and removing `cudatoolkit` from the `buildInputs` but I get the same cmake error.

Reply:
Random guess: could upstream have updated to cuda12 now? In either case, we don't want to ever introduce `cudatoolkit` back in again, so we should be debugging the other error

I just tried building with cuda12, but we need to merge Connor's PR for relaxing the pytorch'es cuda version constraints, and the fix for `<nv/target>`: https://github.com/NixOS/nixpkgs/pull/261656 https://github.com/NixOS/nixpkgs/pull/261653

Reply:
Updated guess: could the upstream python build script have started hiding our `CUDAToolkit_INCLUDE_DIRS` environment variable? Reminder: our current pytorch build doesn't execute CMake directly and, in principle, would happily let the python scripts override our flags with the wrong ones

Reply:
(ideally, we shouldn't rely on CUDAToolkit_INCLUDE_DIRS the environment variable, but pass it directly using `-D`)

Reply:
For me, the issue is that it does not fail because of a specific library lookup error.
It fails as soon as the cmake check error: It seems that it tries to look for cudatoolkit itself...
I guess that it happens [here](https://github.com/pytorch/pytorch/blob/7bd004297a8280af138bbd92d370fcae9aa1b4a9/cmake/Modules/FindCUDAToolkit.cmake#L496-L562)

Reply:
This is where it crashes to be exact: https://github.com/pytorch/pytorch/blob/v2.1.0/cmake/Modules/FindCUDAToolkit.cmake#L837

Reply:
>This is where it crashes to be exact: https://github.com/pytorch/pytorch/blob/v2.1.0/cmake/Modules/FindCUDAToolkit.cmake#L837

Ah, looks like we're missing some CMake variables then?

Reply:
I tried to fix the cudatoolkit-free build and got this:


```
CMake Error at cmake/public/cuda.cmake:65 (message):
  Found two conflicting CUDA installs:

  V11.8.89 in
  '/nix/store/cm0kgn5995f8s6wddp9pgf172cr4280h-cuda_cudart-11.8.89/include'
  and

  V11.8.89 in
  '/nix/store/r5vkgz03d42n7xivv7zfqnm3v4jvfg0v-cuda_cccl-11.8.89-dev/include;/nix/store/cm0kgn5995f8s6wddp9pgf172cr4280h-cuda_cudart-11.8.89/include;/nix/store/iyzbvj3qlq4b99l1r5507ms9hq5cl1d2-cuda_cupti-11.8.87-dev/include;/nix/store/3ygzadn7c72k5i10qffzainzgmd27dpf-cuda_nvcc-11.8.89-dev/include;/nix/store/5h2lnvvdh29mwyaind8s4rgy5fry6ik9-cuda_nvml_dev-11.8.86-dev/include;/nix/store/lqpdlm7q5a2jy5w2qkc2b2hw4048i3zk-cuda_nvrtc-11.8.89-dev/include;/nix/store/i46q6mggmvcddk9hm0d36mxrbvh4di8c-cuda_nvtx-11.8.86-dev/include;/nix/store/7jx76sywhvhy1viwh00y71r3yyamji9d-libcublas-11.11.3.6-dev/include;/nix/store/x1vbhxmkwg62yfn5kvjj16lr3r4fc6n4-libcufft-10.9.0.58-dev/include;/nix/store/1kfria0ian7hrm8cy1xhpzjaw5zcgw1r-libcurand-10.3.0.86-dev/include;/nix/store/wzqs5agmc5mj89267cl24gyab5s31c7n-libcusolver-11.4.1.48-dev/include;/nix/store/l4lq7k62wms1q9y39jv8pv2qsr4zyn99-libcusparse-11.7.5.86-dev/include;/nix/store/zpjpcd8snwr8znx56kqbllpmg990n1hf-cuda_profiler_api-11.8.86-dev/include'
Call Stack (most recent call first):
  cmake/Dependencies.cmake:44 (include)
  CMakeLists.txt:722 (include)
```

Reply:
Ah right, their custom logic for syncing FindCUDA and FindCUDAToolkit afaiu. I thought Connor had to deal with the same stuff when migrating torch to the hook

Reply:
I tried `-DCUDA_INCLUDE_DIRS=$CUDAToolkit_INCLUDE_DIR` but they must be resetting the variable somwhere so I get the same error 😭

Reply:
May be related to https://github.com/pytorch/pytorch/issues/108931#issuecomment-1712397556 (judging by the versions they report), except they have copies of the same directory and we literally have the exact same path, except the legacy FindCUDA stuff only reports the path to cudart, and the FindCUDAToolkit module reports, as it should, the list of all components

Reply:
So, replacing `STREQUAL` with `IN_LIST` does work, but then I encounter an error about [`CUDA::nvToolsExt`](https://github.com/pytorch/pytorch/blob/57c7aa12dbf71617bd21fe7e076df8e823b5b7bb/cmake/public/cuda.cmake#L70-L72) discovery. Idk why do they do that check manually, because they [do not ask for any specific components in `find_package(CUDAToolkit)`](https://github.com/pytorch/pytorch/blob/57c7aa12dbf71617bd21fe7e076df8e823b5b7bb/cmake/public/cuda.cmake#L59)

Reply:
Naive question: why including `cudatoolkit` is that problematic ?
When you get the `whl` (binary) version of torch, it includes a full cuda binary right ?
Why would it be a problem to have that for the source-based version ?
Maybe I missed something...

Reply:
> why including cudatoolkit is that problematic ?

Because `cudatoolkit` is populated from the `.runfile`, which isn't in any way marked as redistributable, and which brings all of cuda components (including the gui tools) in one (two) output(s), and which we don't properly maintain, and which increases the runtime closure by some 10-15GiB of stuff we don't ever use, and changes the `outputPath`s way more frequently because of how many dependencies affect it

Reply:
`cudatoolkit` \approx `symlinkJoin (filter isDerivation (attrValues cudaPackages))` and you can't discard anything from the runtime closure because everything is linked through the single output and legally ambiguous

Reply:
RE: nvToolsExt

Uhmmm, I tried `rm cmake/Modules/FindCUDAToolkit.cmake` and I think now everything works

Reply:
Ok, now nvcc complains about gcc12 because we somehow dropped the `backendStdenv` thing I think

Reply:
Lmao, I made it use backendStdenv and now nvlink fails with undefined references...

Reply:
I'm comparing the upstream `FindCUDAToolkit.cmake` against the vendored noe from pytorch/pytorch, and upstream hasn't got any variables the like of `CUDAToolkit_nvToolsExt_INCLUDE_DIR`

Reply:
I think they did something custom

Reply:
https://github.com/pytorch/pytorch/issues/112857

Reply:
> undefined references

Oh, this is because we've been setting a bunch env variables for the `setup.py` script, and once I switched over to invoking CMake directly torch started building its own nccl from the git submodules...

Reply:
> Naive question: why including `cudatoolkit` is that problematic ? When you get the `whl` (binary) version of torch, it includes a full cuda binary right ? Why would it be a problem to have that for the source-based version ? Maybe I missed something...

Fun fact: they actually have a bunch of scripts which manually unpack things and `rm`/`cp` libraries around to try to include only those CUDA dependencies which are needed before they upload their wheels. You can see the mountain of bash here: https://github.com/pytorch/builder

Reply:
As on Matrix in the CUDA maintainers space, we don't want to include the CUDA Toolkit runfile installer for a number of reasons. Let's see what can be done to get around this build system nonsense :)